### PR TITLE
Add `amp_scaling_state` (apex) migration

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

See title

Fixes https://github.com/Lightning-AI/lightning/pull/16149#discussion_r1054271661

### Does your PR introduce any breaking changes? If yes, please list them.

None

cc @justusschock @awaelchli @carmocca

### CODE REVIEW

File: src/pytorch_lightning/utilities/migration/migration.py [hunk 1]
```diff
@@ -200,3 +201,18 @@ def new_key(old_key: str) -> str:
...
+    """Apex support was removed in v2.0.00, and this migration drops it from the state-keys saved in the checkpoint
+    dict.
+
+    Version: 2.0.0
+    Commit: e544676ff434ed96c6dd3b4e73a708bcb27ebcf1
+    PR: #16149
+    """
+    key = "amp_scaling_state"
+    if key in checkpoint:
+        rank_zero_warn("This checkpoint contains apex AMP data, but apex support has been removed.")
```
Comment (awaelchli) [inline src/pytorch_lightning/utilities/migration/migration.py:216]:
```suggestion
        rank_zero_warn("This checkpoint contains apex AMP data, but apex support has been removed in v2.0.0.")
```

File: src/pytorch_lightning/utilities/migration/migration.py [hunk 2]
```diff
@@ -200,3 +201,18 @@ def new_key(old_key: str) -> str:
 
     checkpoint["callbacks"] = new_callback_states
     return checkpoint
+
+
+def _drop_apex_amp_state_from_checkpoint(checkpoint: _CHECKPOINT) -> _CHECKPOINT:
+    """Apex support was removed in v2.0.00, and this migration drops it from the state-keys saved in the checkpoint
```
Comment (awaelchli) [inline src/pytorch_lightning/utilities/migration/migration.py:207]:
```suggestion
    """Apex support was removed in v2.0.0, and this migration drops it from the state-keys saved in the checkpoint
```

**Review by awaelchli — APPROVED** [review]

File: tests/tests_pytorch/utilities/migration/test_migration.py [hunk 3]
```diff
@@ -149,3 +149,12 @@ def test_migrate_model_checkpoint_save_on_train_epoch_end_default_collision():
     with pytest.warns(PossibleUserWarning, match="callback states in this checkpoint.* colliding with each other"):
         updated_checkpoint, _ = migrate_checkpoint(old_checkpoint.copy())
     assert updated_checkpoint["callbacks"] == old_checkpoint["callbacks"]  # no migration was performed
+
+
+def test_migrate_dropped_amp_state():
```
Comment (carmocca) [inline tests/tests_pytorch/utilities/migration/test_migration.py:154]:
```suggestion
def test_migrate_dropped_apex_amp_state():
```

File: src/pytorch_lightning/utilities/migration/migration.py [hunk 4]
```diff
@@ -47,6 +47,7 @@ def _migration_index() -> Dict[str, List[Callable[[_CHECKPOINT], _CHECKPOINT]]]:
         "1.6.0": [_migrate_loop_global_step_to_progress_tracking, _migrate_loop_current_epoch_to_progress_tracking],
         "1.6.5": [_migrate_loop_batches_that_stepped],
         "1.9.0": [_migrate_model_checkpoint_save_on_train_epoch_end_default],
+        "2.0.0": [_drop_apex_amp_state_from_checkpoint],
```
Comment (awaelchli) [inline src/pytorch_lightning/utilities/migration/migration.py:50]:
```suggestion
        "2.0.0": [_drop_apex_amp_state_from_checkpoint],
```
technically "from_checkpoint" is redundant for any of the migration names

File: src/pytorch_lightning/utilities/migration/migration.py [hunk 5]
```diff
@@ -200,3 +201,18 @@ def new_key(old_key: str) -> str:
 
     checkpoint["callbacks"] = new_callback_states
     return checkpoint
+
+
+def _drop_apex_amp_state_from_checkpoint(checkpoint: _CHECKPOINT) -> _CHECKPOINT:
```
Comment (carmocca) [inline src/pytorch_lightning/utilities/migration/migration.py:206]:
```suggestion
def _drop_apex_amp_state(checkpoint: _CHECKPOINT) -> _CHECKPOINT:
```

File: src/pytorch_lightning/utilities/migration/migration.py (hunk 4 above)
Comment (carmocca) [reply src/pytorch_lightning/utilities/migration/migration.py:50]:
```suggestion
        "2.0.0": [_drop_apex_amp_state],
```

**Review by Borda — APPROVED** [review]

**Pull request merged**